File: LintCode/Easy/93. Balanced Binary Tree/Solution.java
/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */
class ResultType {
    public boolean isBalanced;
    public int maxDepth;
    
    public ResultType(boolean isBalanced, int maxDepth) {
        this.isBalanced = isBalanced;
        this.maxDepth = maxDepth;
    }
}


public class Solution {
    /**
     * @param root: The root of binary tree.
     * @return: True if this Binary tree is Balanced, or false.
     */
    public boolean isBalanced(TreeNode root) {
        // write your code here
        return helper(root).isBalanced;
    }
    
    
    private ResultType helper(TreeNode root) {
        
        if (root == null) {
            return new ResultType(true, 0);
        }

        ResultType left = helper(root.left);
        ResultType right = helper(root.right);

        // subtree is not balanced
        if (!left.isBalanced || !right.isBalanced) {
            return new ResultType(false, -1);
        }
        
        // root is not balanced
        if (Math.abs(left.maxDepth - right.maxDepth) > 1) {
            return new ResultType(false, -1);
        }
        
        // current node is balanced
        return new ResultType(true, Math.max(left.maxDepth, right.maxDepth) + 1);
        
    }
    

    
}

File: LintCode/Easy/93. Balanced Binary Tree/Solution.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""
class ResultType:
    def __init__(self, depth, isBalanced):
        self.depth = depth
        self.isBalanced = isBalanced


class Solution:
    """
    @param root: The root of binary tree.
    @return: True if this Binary tree is Balanced, or false.
    """

    def isBalanced(self, root):
        # write your code here
        if not root:
            return True
        return self.helper(root).isBalanced


    def helper(self, root):
        if not root:
            return ResultType(0, True)

        # Divide -> find left subtree depth and right subtree depth
        left  = self.helper(root.left)
        right = self.helper(root.right)

        # return False if subtree is not balanced
        if not left.isBalanced or not right.isBalanced:
            return ResultType(0, False)

        # return false if current subtree depth difference is greater than 1
        if abs(left.depth - right.depth) > 1:
            return ResultType(0, False)

        return ResultType(max(left.depth, right.depth) + 1, True)
